Code/Framework/AzCore/AzCore/Metrics/EventLoggerFactoryImpl.cpp

```diff
@@ -0,0 +1,139 @@
+ * For complete copyright and license terms please see the LICENSE at the root of this distribution.
+ *
+ * SPDX-License-Identifier: Apache-2.0 OR MIT
+ *
+ */
+
+#include "EventLoggerFactoryImpl.h"
+#include <AzCore/Metrics/IEventLogger.h>
+#include <AzCore/std/functional.h>
+#include <AzCore/std/parallel/scoped_lock.h>
+#include <AzCore/std/ranges/ranges_algorithm.h>
+
+namespace AZ::Metrics
+{
+    EventLoggerFactoryImpl::EventLoggerDeleter::EventLoggerDeleter() = default;
+    EventLoggerFactoryImpl::EventLoggerDeleter::EventLoggerDeleter(bool shouldDelete)
+        : m_delete(shouldDelete)
+    {}
+    void EventLoggerFactoryImpl::EventLoggerDeleter::operator()(IEventLogger* ptr) const
+    {
+        if (m_delete)
+        {
+            delete ptr;
+        }
+    }
+
+    EventLoggerFactoryImpl::EventLoggerFactoryImpl() = default;
+    EventLoggerFactoryImpl::~EventLoggerFactoryImpl() = default;
+
+    void EventLoggerFactoryImpl::VisitEventLoggers(const VisitEventLoggerInterfaceCallback& callback) const
+    {
+        auto VisitInterface = [&callback](const IdToEventLoggerEntry& eventLoggerEntry)
+        {
+            return eventLoggerEntry.m_logger != nullptr ? callback(*eventLoggerEntry.m_logger) : true;
+        };
+
+        AZStd::scoped_lock lock(m_eventLoggerMutex);
+        AZStd::ranges::all_of(m_eventLoggers, VisitInterface);
+    }
+
+    AZ::Outcome<void, AZStd::unique_ptr<IEventLogger>> EventLoggerFactoryImpl::RegisterEventLogger(EventLoggerId eventLoggerId, AZStd::unique_ptr<IEventLogger> eventLogger)
+    {
+        // Transfer ownership to a temporary EventLoggerPtr which is supplied to the RegisterEventLoggerImpl
+        // If registration fails, the event logger pointer is returned in the failure outcome
+        if (auto registerResult = RegisterEventLoggerImpl(eventLoggerId, EventLoggerPtr{ eventLogger.release() });
+            !registerResult.IsSuccess())
+        {
+            return AZ::Failure(AZStd::unique_ptr<IEventLogger>(registerResult.TakeError().release()));
+        }
+
+        // registration succeeded, return a void success outcome
+        return AZ::Success();
+    }
+
+    bool EventLoggerFactoryImpl::RegisterEventLogger(EventLoggerId eventLoggerId, IEventLogger& eventLogger)
+    {
+        // Create a temporary EventLoggerPtr with a custom deleter that does NOT delete the eventLogger reference
+        // On success, the the EventLoggerPtr is stored in the event logger array and will not the reference to the input eventLogger
+        // when the eventLogger is unregistered
+        return RegisterEventLoggerImpl(eventLoggerId, EventLoggerPtr{ &eventLogger, EventLoggerDeleter{false} }).IsSuccess();
+    }
+
+    auto EventLoggerFactoryImpl::RegisterEventLoggerImpl(EventLoggerId eventLoggerId, EventLoggerPtr eventLoggerPtr)
+        -> AZ::Outcome<void, EventLoggerPtr>
+    {
+        if (eventLoggerPtr == nullptr)
+        {
+            return AZ::Failure(AZStd::move(eventLoggerPtr));
```

Comment:
q: Why is a AZStd::move necessary if eventLoggerPtr is ```nullptr```?

Reply:
Because the `EventLoggerPtr` type is not being returned, but an `AZ::Outcome<void, EventLoggerPtr>` type itself.
So a move is needed to construct the Outcome with the `EventLoggerPtr` which is a type to an `AZStd::unique_ptr<IEventLogger>`